node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupIndices.java

```diff
@@ -0,0 +1,163 @@
+    }
+
+    /** Unretire nodes that are already in the correct group when the group is deficient. */
+    private NodeCandidate unretireNodeInExpandedGroup(NodeCandidate node, int[] countInGroup) {
+        if ( ! node.allocation().get().membership().retired()) return node;
+        var currentGroup = node.allocation().get().membership().cluster().group();
+        if (currentGroup.isEmpty()) return node;
+        if (currentGroup.get().index() >= requested.groups()) return node;
+        if (node.preferToRetire() || node.wantToRetire()) return node;
+        if (requested.count().isPresent() && countInGroup[currentGroup.get().index()] >= requested.count().get() / requested.groups()) return node;
+        node = unretire(node);
+        if (node.allocation().get().membership().retired()) return node;
+        countInGroup[currentGroup.get().index()]++;
+        return node;
+    }
+
+    private NodeCandidate inFirstGroupWithDeficiency(NodeCandidate node, int[] countInGroup) {
+        for (int group = 0; group < requested.groups(); group++) {
+            if (requested.count().isEmpty() || countInGroup[group] < requested.count().get() / requested.groups()) {
+                return inGroup(group, node, countInGroup);
+            }
+        }
+        return node;
+    }
+
+    private boolean shouldRemove(NodeCandidate node) {
+        var currentGroup = node.allocation().get().membership().cluster().group();
+        if (currentGroup.isEmpty()) return true; // new and not assigned an index: Not needed
+        return currentGroup.get().index() >= requested.groups();
+    }
+
+    private NodeCandidate inGroup(int group, NodeCandidate node, int[] countInGroup) {
+        node = unretire(node);
+        if (node.allocation().get().membership().retired()) return node;
+        var membership = node.allocation().get().membership();
+        var currentGroup = membership.cluster().group();
+        countInGroup[group]++;
+        if ( ! currentGroup.isEmpty() && currentGroup.get().index() < requested.groups())
+            countInGroup[membership.cluster().group().get().index()]--;
+        return node.withNode(node.toNode().with(node.allocation().get().with(membership.with(membership.cluster().with(Optional.of(ClusterSpec.Group.from(group)))))));
+    }
+
+    /** Attempt to unretire the given node if it is retired. */
+    private NodeCandidate unretire(NodeCandidate node) {
+        if (node.retiredNow()) return node;
+        if ( ! node.allocation().get().membership().retired()) return node;
+        if ( ! hasCompatibleResources(node) ) return node;
+        var parent = node.parentHostname().flatMap(hostname -> allNodes.node(hostname));
+        if (parent.isPresent() && (parent.get().status().wantToRetire() || parent.get().status().preferToRetire())) return node;
+        node = node.withNode();
+        if ( ! requested.isCompatible(node.resources()))
```

Comment:
Isn't it necessary to check the free parent capacity here, as well?

Reply:
I see that wasn't done before, so could do separately, if needed.

Reply:
We don't create candidates on hosts where there isn't capacity.

Reply:
Ah, that's good.